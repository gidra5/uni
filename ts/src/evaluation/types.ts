import { AbstractSyntaxTree } from "../parser/ast";
import { Scope } from "../scope";

export type SymbolValue = symbol;
export type ScopeValue = { get?: () => Value; set?: (val: Value) => void };
export type ExprValue = {
  kind: "expr";
  ast: AbstractSyntaxTree;
  scope: Scope<ScopeValue>;
};
export type RecordValue = {
  kind: "record";
  get: (key: Value) => Value;
  set: (key: Value, val: Value) => void;
  has: (key: Value) => boolean;
  tuple: Value[];
  record: Record<string | symbol, Value>;
  map: Map<Value, Value>;
};
export type TypeValue = { kind: "type"; name: string; value: Value };
export type FunctionValue = (argChannel: symbol) => symbol;
export type Value =
  | number
  | string
  | boolean
  | null
  | ExprValue
  | FunctionValue
  | RecordValue
  | SymbolValue
  | TypeValue;
export type TaskQueueContext = { scope: Scope<ScopeValue> };
