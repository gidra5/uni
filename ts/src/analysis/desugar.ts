import { node, NodeType, Tree } from "../ast";
import { assert } from "../utils";

export const desugar = (ast: Tree): Tree => {
  ast.children = ast.children.map(desugar);
  switch (ast.type) {
    case NodeType.PARENS:
      assert(ast.children.length === 1);
      if (ast.children[0].type === NodeType.IMPLICIT_PLACEHOLDER) return node(NodeType.UNIT);
      return ast.children[0];
    case NodeType.SCRIPT:
    case NodeType.MODULE:
    default:
      return ast;
  }
};
