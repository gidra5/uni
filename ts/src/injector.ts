import { FileMap } from "codespan-napi";
import type { Precedence } from "./ast";
import type { Position } from "./position";

enum Injectable {
  FileMap = "FileMap",
  RootDir = "RootDir",
  ASTNodeNextId = "ASTNodeNextId",
  ASTNodePrecedenceMap = "ASTNodePrecedenceMap",
  ASTNodePositionMap = "ASTNodePositionMap",
}

type InjectableType = {
  [Injectable.FileMap]: FileMap;
  [Injectable.RootDir]: string;
  [Injectable.ASTNodeNextId]: number;
  [Injectable.ASTNodePrecedenceMap]: Map<string, Precedence>;
  [Injectable.ASTNodePositionMap]: Map<string, Position>;
};

const registry = new Map<string, any>();

const register = <const T extends Injectable>(name: T, value: InjectableType[T]) => {
  registry.set(name, value);
};

const inject = <const T extends Injectable>(name: T): InjectableType[T] => {
  if (!registry.has(name)) {
    throw new Error(`Missing injection entry for: ${name}`);
  }

  return registry.get(name);
};

// Register default injectables
register(Injectable.RootDir, process.cwd());
register(Injectable.FileMap, new FileMap());
register(Injectable.ASTNodeNextId, 0);
register(Injectable.ASTNodePrecedenceMap, new Map());
register(Injectable.ASTNodePositionMap, new Map());

export { register, inject, Injectable };
