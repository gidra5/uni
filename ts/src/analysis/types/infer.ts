import { Iterator } from "iterator-js";
import { NodeType, Tree } from "../../ast";
import { assert, nextId, unreachable } from "../../utils";
import { inject, Injectable } from "../../utils/injector";
import { UnificationTable } from "./unification";
import { PhysicalType, Type } from "./utils";
import { Binding, resolve } from "../scope";

export type TypeSchema = Map<number, Type>;
export type PhysicalTypeSchema = Map<number, PhysicalType>;

// subtype infers types that values must always satisfy (the actual values will be at least of these types)
// for example, if a parameter is only used as int (passed to a function that only accepts ints),
// then the type for the parameter will be int
// its value must inhabit the type
// if nothing can be inferred, then the type will be "unknown"

// supertype infers types that values will always satisfy (the actual values will be at most of these types)
// for example, if a function only called with ints, then the type for its arguments will be int

const scope = [
  ["print_int", { fn: { arg: "int", return: "int" } }],
  ["print_string", { fn: { arg: "string", return: "string" } }],
  ["print_float", { fn: { arg: "float", return: "float" } }],
  ["print_symbol", { fn: { arg: "symbol", return: "symbol" } }],
  ["true", "boolean"],
  ["false", "boolean"],
] as const satisfies [string, Type][];

export const globalResolvedNames = Iterator.iter(scope)
  .map<Binding>(([name]) => [name, nextId()])
  .toMap();
export const globalResolvedNamesArray = [...globalResolvedNames.entries()];

const globalNames = new Map<number, Type>(scope.map(([name, type]) => [globalResolvedNames.get(name)!, type]));

export class Context {
  unificationTable = new UnificationTable();
  names: Map<number, Type> = globalNames;
  constructor() {}

  bind(ast: Tree, f: (argType: Type) => Type): Type {
    const argType = inferPattern(ast);
    const name = inject(Injectable.NodeToVariableMap).get(ast.children[0].id)!;
    this.names.set(name, argType);
    const t = f(argType);
    this.names.delete(name);
    return t;
  }
}

// is `a` a subtype of `b`? `a <: b` == true
// the bottom type must always be a subtype of the top type
export const isSubtype = (a: Type, b: Type): boolean => {
  // console.log(a, b);
  if (a === b) return true;

  if (typeof b === "object" && "not" in b) return !isSubtype(b.not, a);
  if (typeof a === "object" && "not" in a) return !isSubtype(b, a.not);

  if (a === "void") return true;
  if (b === "unknown") return true;

  if (typeof a === "object" && "and" in a) {
    return a.and.some((type) => isSubtype(type, b));
    // // console.log(a, b);
    // const x = a.and.map((type) => isSubtype(type, b));
    // // console.log(3, x);
    // return x.some((x) => x);
  }
  if (typeof b === "object" && "and" in b) {
    return b.and.every((type) => isSubtype(a, type));
    // // console.log(a, b);
    // const x = b.and.map((type) => isSubtype(a, type));
    // // console.log(4, x);
    // return x.every((x) => x);
  }

  if (typeof a === "object" && "or" in a) {
    return a.or.every((type) => isSubtype(type, b));
    // console.log(a, b);
    // const x = a.or.map((type) => isSubtype(type, b));
    // console.log(2, x);
    // return x.every((x) => x);
  }
  if (typeof b === "object" && "or" in b) {
    return b.or.some((type) => isSubtype(a, type));
    // console.log(a, b);
    // const x = b.or.map((type) => isSubtype(a, type));
    // console.log(1, x);

    // return x.some((x) => x);
  }

  if (typeof a === "object" && "fn" in a) {
    if (!(typeof b === "object" && "fn" in b)) return false;
    return isSubtype(b.fn.arg, a.fn.arg) && isSubtype(a.fn.return, b.fn.return);
  }
  if (typeof b === "object" && "fn" in b) return false;

  if (typeof a === "object" && "variable" in a) {
    if (typeof b === "object" && "variable" in b) return a.variable === b.variable;
    return false;
  }
  if (typeof b === "object" && "variable" in b) return false;

  if (typeof a === "object" && "record" in a) {
    if (!(typeof b === "object" && "record" in b)) return false;
    if (b.record.length === 0) return true;
    if (b.record.length > a.record.length) return false;
    if (a.labels) {
      if (!b.labels) return false;
      const bMap = Iterator.iter(b.labels).zip(b.record).toMap();
      return Iterator.iter(a.labels)
        .zip(a.record)
        .every(([label, type]) => bMap.has(label) && isSubtype(type, bMap.get(label)!));
    }
    return Iterator.iter(a.record)
      .zip(b.record)
      .every(([a, b]) => isSubtype(a, b));
  }

  return false;
};

export const isSubtypeEqual = (a: Type, b: Type): boolean => isSubtype(a, b) && isSubtype(b, a);

const inferPattern = (a: Tree): Type => {
  switch (a.type) {
    case NodeType.NAME:
      if (a.data.value === "void") return "void";
      if (a.data.value === "unknown") return "unknown";
      if (a.data.value === "int") return "int";
      if (a.data.value === "float") return "float";
      if (a.data.value === "string") return "string";
      if (a.data.value === "bool") return "boolean";
      return { variable: a.id };
    case NodeType.LABEL:
      return inferPattern(a.children[1]);
    case NodeType.FUNCTION:
      return { fn: { arg: inferPattern(a.children[0]), return: inferPattern(a.children[1]) } };
    default:
      unreachable("cant infer type");
  }
};

export const infer = (ast: Tree, context: Context): Type => {
  console.dir({ log: 1, context, ast }, { depth: null });

  switch (ast.type) {
    case NodeType.NUMBER: {
      const type = Number.isInteger(ast.data.value) ? "int" : "float";
      context.unificationTable.addConstraint(ast.id, { exactly: type });
      return type;
    }
    case NodeType.STRING:
      context.unificationTable.addConstraint(ast.id, { exactly: "string" });
      return "string";
    case NodeType.NAME: {
      const name = inject(Injectable.NodeToVariableMap).get(ast.id)!;
      const type = context.names.get(name);

      if (type) context.unificationTable.addConstraint(ast.id, { exactly: type });
      else context.names.set(name, { variable: ast.id });

      return type ?? { variable: ast.id };
    }
    case NodeType.SCRIPT:
    case NodeType.SEQUENCE: {
      const types = ast.children.map((child) => infer(child, context));
      context.unificationTable.addConstraint(ast.id, { equals: ast.children[types.length - 1].id });
      return types[types.length - 1];
    }
    case NodeType.DELIMITED_APPLICATION:
    case NodeType.APPLICATION: {
      infer(ast.children[0], context);
      const arg_type = infer(ast.children[1], context);
      const return_type: Type = { variable: ast.id };

      context.unificationTable.addConstraint(ast.id, { exactly: return_type });
      context.unificationTable.addConstraint(ast.children[0].id, {
        exactly: { fn: { arg: arg_type, return: return_type } },
      });

      return return_type;
    }
    case NodeType.FUNCTION: {
      return context.bind(ast.children[0], (argType) => {
        const returnType = infer(ast.children[1], context);
        const type = { fn: { arg: argType, return: returnType } };
        context.unificationTable.addConstraint(ast.id, { exactly: type });
        return type;
      });
    }
    case NodeType.ADD: {
      context.unificationTable.addConstraint(ast.id, { exactly: "int" });
      return "int";
    }
    case NodeType.MODULE:
    default:
      unreachable(`cant infer top types ${ast.type}`);
  }
};

const constrain = (ast: Tree, context: Context, expectedType: Type): void => {};

export const inferPhysical = (typeMap: TypeSchema): PhysicalTypeSchema => {
  const physicalTypeMap: PhysicalTypeSchema = new Map();

  const translate = (type: Type, id?: number): PhysicalType => {
    if (type === "int") return { int: 32 };
    if (type === "float") return { float: 32 };
    if (type === "string") return { pointer: { int: 8 } };
    if (type === "void") return "void";
    if (type === "unknown") return "unknown";
    assert(typeof type === "object");

    // must be resolved by that point
    assert(!("variable" in type));
    assert(!("and" in type));

    if ("fn" in type) {
      const arg = translate(type.fn.arg);
      const returnType = translate(type.fn.return);
      const freeVars = (id && inject(Injectable.ClosureVariablesMap).get(id)) || [];
      assert(freeVars);
      const closure = freeVars.map((name) => {
        const type = typeMap.get(name)!;
        const physicalType = translate(type, name);
        physicalTypeMap.set(name, physicalType);
        return physicalType;
      });

      return { fn: { args: [arg], ret: returnType, closure } };
    }
    if ("atom" in type) return { int: 32 };

    unreachable("cant convert type to LLVM type");
  };

  typeMap.forEach((type, id) => {
    if (physicalTypeMap.has(id)) return;
    physicalTypeMap.set(id, translate(type, id));
  });

  return physicalTypeMap;
};

export const substituteConstraints = (ast: Tree, context: Context, map: TypeSchema = new Map()): TypeSchema => {
  if (map.has(ast.id)) return map;
  switch (ast.type) {
    case NodeType.FUNCTION: {
      substituteConstraints(ast.children[1], context, map);
      map.set(ast.id, context.unificationTable.resolveType(ast.id));
      return map;
    }
    default: {
      ast.children.forEach((child) => substituteConstraints(child, context, map));
      map.set(ast.id, context.unificationTable.resolveType(ast.id));
      return map;
    }
  }
};

export const inferTypes = (ast: Tree): TypeSchema => {
  const context = new Context();
  resolve(ast, globalResolvedNamesArray);
  infer(ast, context);
  context.unificationTable.truncateTautologies();
  return substituteConstraints(ast, context);
  // console.dir(
  //   {
  //     // ast,
  //     table: context.unificationTable,
  //     // types: inject(Injectable.TypeMap),
  //     closures: inject(Injectable.ClosureVariablesMap),
  //     bound: inject(Injectable.BoundVariablesMap),
  //     nodeToVariable: inject(Injectable.NodeToVariableMap),
  //   },
  //   { depth: null }
  // );
  // inferPhysical();
};
