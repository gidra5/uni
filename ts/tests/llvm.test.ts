import { beforeEach, describe, expect } from "vitest";
import { test } from "@fast-check/vitest";
import { parseTokenGroups } from "../src/parser/tokenGroups";
import { parseScript } from "../src/parser/parser";
import { generateLLVMCode } from "../src/codegen/llvm";
import { exec } from "child_process";
import { Injectable, register } from "../src/utils/injector";
import { FileMap } from "codespan-napi";
import { inferTypes } from "../src/analysis/types/infer";
import { desugar } from "../src/analysis/desugar";

beforeEach(() => {
  register(Injectable.FileMap, new FileMap());
  register(Injectable.PrecedenceMap, new Map());
  register(Injectable.NextId, 0);
  register(Injectable.PositionMap, new Map());
  register(Injectable.TypeMap, new Map());
  register(Injectable.PhysicalTypeMap, new Map());
  register(Injectable.ClosureVariablesMap, new Map());
  register(Injectable.BoundVariablesMap, new Map());
  register(Injectable.NodeToVariableMap, new Map());
});

const testCase = async (src: string) => {
  const tokens = parseTokenGroups(src);
  const ast = parseScript(tokens);
  const desugared = desugar(ast);
  inferTypes(desugared);
  const compiled = generateLLVMCode(desugared);
  expect(compiled).toMatchSnapshot("compiled");

  // const optimized: any[] = [];
  const stdout: any[] = [];
  const stderr: any[] = [];
  const interpreter = exec("lli-18");
  const optimizer = exec(`opt-18 -f -S -O3`);
  interpreter.stdout?.on("data", (data) => stdout.push(data));
  interpreter.stderr?.on("data", (data) => stderr.push(data));
  optimizer.stderr?.on("data", (data) => stderr.push(data));
  optimizer.stdout?.pipe(interpreter.stdin!);
  // optimizer.stdout?.on("data", (data) => optimized.push(data));
  optimizer.stdin?.write(compiled);
  optimizer.stdin?.end();

  await new Promise((resolve) => interpreter.on("exit", resolve));

  // expect(optimized.join("")).toMatchSnapshot("optimized");
  expect(stdout).toMatchSnapshot("stdout");
  expect(stderr).toMatchSnapshot("stderr");
};

describe("compilation", () => {
  test.todo(
    "either",
    async () =>
      await testCase(`print(
        (((fn x -> fn m -> fn n -> m x) 1) (fn x -> x)) (fn x -> x)
      )`)
  );
  test.todo(
    "apply",
    async () =>
      await testCase(`print(
        ((fn f -> fn x -> f x) fn x -> x) 2
      )`)
  );
  test.todo(
    "wrapper",
    async () =>
      await testCase(`print(
        ((fn x -> fn m -> m x) 2) fn x -> x  
      )`)
  );
  test.todo(
    "church tuple",
    async () =>
      await testCase(`print(
        ((fn x -> fn y -> fn m -> m x y) 1 2) fn x -> fn _ -> x
      )`)
  );
  test.todo("function closure", async () => await testCase(`print((fn x -> fn y -> y + 2 * x) 1 2)`));
  test.todo("function deep closure", async () => await testCase(`print((fn x -> fn y -> fn z -> x + y + z) 1 3 5)`));
  test.only("function application and literal print", async () => await testCase(`print((fn x -> x + x) 2)`));
  test("print number", async () => await testCase(`print 1`));
  test("hello world", async () => await testCase(`print "hello world!"`));
  test("hello world twice", async () => await testCase(`print "hello world!"; print "hello world!"`));
  test("two prints", async () => await testCase(`print "hello world!"; print "hello world 2!"`));
  test("hello world string", async () => await testCase(`"hello world!"`));
});
