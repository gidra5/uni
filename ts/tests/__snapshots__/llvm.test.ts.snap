// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compilation > function application and literal print > compiled 1`] = `
"@const_56 = constant [4 x i8] c"%i
\\00"
@const_61 = constant i32 2
define i32 @main() {
  main: 
    %var_62 = load i32, ptr @const_61
    %var_63 = call i32 @fn_58(i32 %var_62)
    %var_64 = call i32 @printInt(i32 %var_63)
    ret i32 0
}
declare i32 @printf(i8*, ...)
define i32 @printInt(i32 %var_55) {
  entry: 
    %var_57 = call i32 @printf(i8* @const_56, i32 %var_55)
    ret i32 %var_57
}
define i32 @fn_58(i32 %var_59) {
  entry: 
    %var_60 = add i32 %var_59, %var_59
    ret i32 %var_60
}
"
`;

exports[`compilation > function application and literal print > stderr 1`] = `[]`;

exports[`compilation > function application and literal print > stdout 1`] = `
[
  "4
",
]
`;

exports[`compilation > function closure > compiled 1`] = `
"@const_74 = constant [4 x i8] c"%i
\\00"
@const_85 = constant i32 2
@const_94 = constant i32 1
define i32 @main() {
  main: 
    %var_95 = load i32, ptr @const_94
    %var_96 = alloca { { i32 }, i32 ({ i32 }, i32)* }
    call void @fn_76({ { i32 }, i32 ({ i32 }, i32)* }* %var_96, i32 %var_95)
    %var_97 = load { { i32 }, i32 ({ i32 }, i32)* }, ptr %var_96
    %var_98 = extractvalue { { i32 }, i32 ({ i32 }, i32)* } %var_97, 0
    %var_99 = extractvalue { { i32 }, i32 ({ i32 }, i32)* } %var_97, 1
    %var_100 = load i32, ptr @const_85
    %var_101 = call i32 %var_99({ i32 } %var_98, i32 %var_100)
    %var_102 = call i32 @printInt(i32 %var_101)
    ret i32 0
}
declare i32 @printf(i8*, ...)
define i32 @printInt(i32 %var_73) {
  entry: 
    %var_75 = call i32 @printf(i8* @const_74, i32 %var_73)
    ret i32 %var_75
}
define void @fn_76(ptr sret({ { i32 }, i32 ({ i32 }, i32)* }) %var_93, i32 %var_77) {
  entry: 
    %var_79 = alloca { i32 }
    %var_80 = load { i32 }, ptr %var_79
    %var_81 = insertvalue { i32 } %var_80, i32 %var_77, 0
    store { i32 } %var_81, ptr %var_79
    %var_89 = alloca { { i32 }, i32 ({ i32 }, i32)* }
    %var_90 = load { { i32 }, i32 ({ i32 }, i32)* }, ptr %var_89
    %var_91 = insertvalue { { i32 }, i32 ({ i32 }, i32)* } %var_90, { i32 } %var_81, 0
    %var_92 = insertvalue { { i32 }, i32 ({ i32 }, i32)* } %var_91, i32 ({ i32 }, i32)* @fn_78, 1
    store { { i32 }, i32 ({ i32 }, i32)* } %var_92, ptr %var_89
    store { { i32 }, i32 ({ i32 }, i32)* } %var_92, ptr %var_93
    ret void
}
define i32 @fn_78({ i32 } %var_82, i32 %var_83) {
  entry: 
    %var_84 = extractvalue { i32 } %var_82, 0
    %var_86 = load i32, ptr @const_85
    %var_87 = mul i32 %var_86, %var_84
    %var_88 = add i32 %var_83, %var_87
    ret i32 %var_88
}
"
`;

exports[`compilation > function closure > stderr 1`] = `[]`;

exports[`compilation > function closure > stdout 1`] = `
[
  "4
",
]
`;

exports[`compilation > function deep closure > compiled 1`] = `
"@const_91 = constant [4 x i8] c"%i
\\00"
@const_123 = constant i32 1
@const_127 = constant i32 3
@const_133 = constant i32 5
define i32 @main() {
  main: 
    %var_124 = load i32, ptr @const_123
    %var_125 = alloca { { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* }
    call void @fn_93({ { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* }* %var_125, i32 %var_124)
    %var_126 = load { { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* }, ptr %var_125
    %var_128 = extractvalue { { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* } %var_126, 0
    %var_129 = extractvalue { { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* } %var_126, 1
    %var_130 = load i32, ptr @const_127
    %var_131 = alloca { { i32, i32 }, i32 ({ i32, i32 }, i32)* }
    call void %var_129({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }* %var_131, { i32 } %var_128, i32 %var_130)
    %var_132 = load { { i32, i32 }, i32 ({ i32, i32 }, i32)* }, ptr %var_131
    %var_134 = extractvalue { { i32, i32 }, i32 ({ i32, i32 }, i32)* } %var_132, 0
    %var_135 = extractvalue { { i32, i32 }, i32 ({ i32, i32 }, i32)* } %var_132, 1
    %var_136 = load i32, ptr @const_133
    %var_137 = call i32 %var_135({ i32, i32 } %var_134, i32 %var_136)
    %var_138 = call i32 @printInt(i32 %var_137)
    ret i32 0
}
declare i32 @printf(i8*, ...)
define i32 @printInt(i32 %var_90) {
  entry: 
    %var_92 = call i32 @printf(i8* @const_91, i32 %var_90)
    ret i32 %var_92
}
define void @fn_93(ptr sret({ { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* }) %var_122, i32 %var_94) {
  entry: 
    %var_96 = alloca { i32 }
    %var_97 = load { i32 }, ptr %var_96
    %var_98 = insertvalue { i32 } %var_97, i32 %var_94, 0
    store { i32 } %var_98, ptr %var_96
    %var_118 = alloca { { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* }
    %var_119 = load { { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* }, ptr %var_118
    %var_120 = insertvalue { { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* } %var_119, { i32 } %var_98, 0
    %var_121 = insertvalue { { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* } %var_120, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* @fn_95, 1
    store { { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* } %var_121, ptr %var_118
    store { { i32 }, void ({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }*, { i32 }, i32)* } %var_121, ptr %var_122
    ret void
}
define void @fn_95(ptr sret({ { i32, i32 }, i32 ({ i32, i32 }, i32)* }) %var_117, { i32 } %var_99, i32 %var_100) {
  entry: 
    %var_101 = extractvalue { i32 } %var_99, 0
    %var_103 = alloca { i32, i32 }
    %var_104 = load { i32, i32 }, ptr %var_103
    %var_105 = insertvalue { i32, i32 } %var_104, i32 %var_101, 0
    %var_106 = insertvalue { i32, i32 } %var_105, i32 %var_100, 1
    store { i32, i32 } %var_106, ptr %var_103
    %var_113 = alloca { { i32, i32 }, i32 ({ i32, i32 }, i32)* }
    %var_114 = load { { i32, i32 }, i32 ({ i32, i32 }, i32)* }, ptr %var_113
    %var_115 = insertvalue { { i32, i32 }, i32 ({ i32, i32 }, i32)* } %var_114, { i32, i32 } %var_106, 0
    %var_116 = insertvalue { { i32, i32 }, i32 ({ i32, i32 }, i32)* } %var_115, i32 ({ i32, i32 }, i32)* @fn_102, 1
    store { { i32, i32 }, i32 ({ i32, i32 }, i32)* } %var_116, ptr %var_113
    store { { i32, i32 }, i32 ({ i32, i32 }, i32)* } %var_116, ptr %var_117
    ret void
}
define i32 @fn_102({ i32, i32 } %var_107, i32 %var_108) {
  entry: 
    %var_109 = extractvalue { i32, i32 } %var_107, 0
    %var_110 = extractvalue { i32, i32 } %var_107, 1
    %var_111 = add i32 %var_109, %var_110
    %var_112 = add i32 %var_111, %var_108
    ret i32 %var_112
}
"
`;

exports[`compilation > function deep closure > stderr 1`] = `[]`;

exports[`compilation > function deep closure > stdout 1`] = `
[
  "9
",
]
`;

exports[`compilation > hello world > compiled 1`] = `
"@const_12 = constant [4 x i8] c"%i
\\00"
@const_14 = constant [13 x i8] c"hello world!\\00"
define i32 @main() {
  main: 
    %var_15 = call i32 @printf(i8* @const_14)
    ret i32 0
}
declare i32 @printf(i8*, ...)
define i32 @printInt(i32 %var_11) {
  entry: 
    %var_13 = call i32 @printf(i8* @const_12, i32 %var_11)
    ret i32 %var_13
}
"
`;

exports[`compilation > hello world > stderr 1`] = `[]`;

exports[`compilation > hello world > stdout 1`] = `
[
  "hello world!",
]
`;

exports[`compilation > hello world string > compiled 1`] = `
"@const_6 = constant [13 x i8] c"hello world!\\00"
define i32 @main() {
  main: 
    ret i32 0
}
"
`;

exports[`compilation > hello world string > stderr 1`] = `[]`;

exports[`compilation > hello world string > stdout 1`] = `[]`;

exports[`compilation > hello world twice > compiled 1`] = `
"@const_26 = constant [4 x i8] c"%i
\\00"
@const_28 = constant [13 x i8] c"hello world!\\00"
define i32 @main() {
  main: 
    %var_29 = call i32 @printf(i8* @const_28)
    %var_30 = call i32 @printf(i8* @const_28)
    ret i32 0
}
declare i32 @printf(i8*, ...)
define i32 @printInt(i32 %var_25) {
  entry: 
    %var_27 = call i32 @printf(i8* @const_26, i32 %var_25)
    ret i32 %var_27
}
"
`;

exports[`compilation > hello world twice > stderr 1`] = `[]`;

exports[`compilation > hello world twice > stdout 1`] = `
[
  "hello world!hello world!",
]
`;

exports[`compilation > print number > compiled 1`] = `
"@const_10 = constant [4 x i8] c"%i
\\00"
@const_12 = constant i32 1
define i32 @main() {
  main: 
    %var_13 = load i32, ptr @const_12
    %var_14 = call i32 @printInt(i32 %var_13)
    ret i32 0
}
declare i32 @printf(i8*, ...)
define i32 @printInt(i32 %var_9) {
  entry: 
    %var_11 = call i32 @printf(i8* @const_10, i32 %var_9)
    ret i32 %var_11
}
"
`;

exports[`compilation > print number > stderr 1`] = `[]`;

exports[`compilation > print number > stdout 1`] = `
[
  "1
",
]
`;

exports[`compilation > two prints > compiled 1`] = `
"@const_26 = constant [4 x i8] c"%i
\\00"
@const_28 = constant [13 x i8] c"hello world!\\00"
@const_30 = constant [15 x i8] c"hello world 2!\\00"
define i32 @main() {
  main: 
    %var_29 = call i32 @printf(i8* @const_28)
    %var_31 = call i32 @printf(i8* @const_30)
    ret i32 0
}
declare i32 @printf(i8*, ...)
define i32 @printInt(i32 %var_25) {
  entry: 
    %var_27 = call i32 @printf(i8* @const_26, i32 %var_25)
    ret i32 %var_27
}
"
`;

exports[`compilation > two prints > stderr 1`] = `[]`;

exports[`compilation > two prints > stdout 1`] = `
[
  "hello world!hello world 2!",
]
`;
